src/sentry/tasks/on_demand_metrics.py

```diff
@@ -0,0 +1,273 @@
+        amount=currently_processing_batch,  # Helps correlate which batch is currently being processed with metrics
+        sample_rate=1.0,
+    )
+    metrics.incr(
+        "task.on_demand_metrics.widgets.pre_rollout.total",
+        amount=dashboard_widget_pre_rollout_count,
+        sample_rate=1.0,
+    )
+    metrics.incr(
+        "task.on_demand_metrics.widgets.total",
+        amount=dashboard_widget_count,
+        sample_rate=1.0,
+    )
+
+
+@instrumented_task(
+    name="sentry.tasks.on_demand_metrics.process_widget_specs",
+    queue="on_demand_metrics",
+    max_retries=0,
+    soft_time_limit=60,
+    time_limit=120,
+    expires=60,
+)
+def process_widget_specs(widget_query_ids: List[int], *args, **kwargs) -> None:
+    """
+    Child task spawned from :func:`schedule_on_demand_check`.
+    """
+    if not options.get("on_demand_metrics.check_widgets.enable"):
+        return
+
+    widget_query_count = 0
+    widget_query_high_cardinality_count = 0
+
+    for query in DashboardWidgetQuery.objects.filter(id__in=widget_query_ids).select_related(
+        "widget__dashboard__organization"
+    ):
+        widget_query_count += 1
+
+        # This only exists to make sure we're 1:1 with flagr since we're not fully rolled out.
+        # TODO: Remove feature flag check once we're check metrics have gone to 0.
+        organization = query.widget.dashboard.organization
+        enabled_features = on_demand_metrics_feature_flags(organization)
+        if "organizations:on-demand-metrics-extraction-widgets" not in enabled_features:
+            metrics.incr(
+                "task.on_demand_metrics.widget_queries.per_run.flag_disabled",
+                sample_rate=1.0,
+            )
+            continue
+
+        is_low_cardinality = _set_widget_query_low_cardinality(query, organization)
+        if not is_low_cardinality:
+            widget_query_high_cardinality_count += 1
+
+    metrics.incr(
+        "task.on_demand_metrics.widget_queries.per_run.high_cardinality",
+        amount=widget_query_high_cardinality_count,
+        sample_rate=1.0,
+    )
+    metrics.incr(
+        "task.on_demand_metrics.widget_queries.per_run.total",
+        amount=widget_query_count,
+        sample_rate=1.0,
+    )
+
+
+def _set_widget_query_low_cardinality(
+    widget_query: DashboardWidgetQuery, organization: Organization
+):
+    """
+    Checks cardinality of existing widget queries before allowing the metric spec, so that
+    group by clauses with high-cardinality tags are not added to the on_demand metric.
+
+    New queries will be checked upon creation and not allowed at that time.
+    """
+
+    max_cardinality_allowed = options.get("on_demand.max_widget_cardinality.count")
+    cache_key = _get_widget_query_cardinality_cache_key(widget_query)
+
+    # We default low cardinality to true since if it's false we'll remove user data.
+    is_low_cardinality = cache.get(cache_key, default=True)
```

Comment:
Note to others: This is the main fix for this PR as with the current set-up we lose specs from being generated when the Soft Time limit is exceeded for the relay task.